rfc/rfc-0000-input-output-correlation.md

```diff
@@ -0,0 +1,217 @@
+# Meta
+
+[meta]: #meta
+
+- Name: Input-Output correlation
+- Start Date: 2022-04-07
+- Author(s): idoru
+- Status: Draft
+- RFC Pull Request: (leave blank)
+- Supersedes:
+  [Draft RFC Read Resources Only When In Success State](https://github.com/vmware-tanzu/cartographer/pull/556)
+- Depends on: [Allow Resources to Report Status](https://github.com/vmware-tanzu/cartographer/pull/738)
+
+# Summary
+
+[summary]: #summary
+
+Cartographer may stamp out arbitrary resources in servicing a `Workload` or `Deliverable`. These resources produce
+output artifacts, and in, turn, may also consume, as inputs, the outputs from other resources.
+
+As of this writing, there is no mechanism to assert which input artifacts are implicated for a given output.
+
+This RFC introduces mechanism which allows template authors to provide additional context on the stamped resources which
+Cartographer can use to make such assertions possible.
+
+It borrows heavily on prior art
+in [Draft RFC Read Resources Only When In Success State](https://github.com/vmware-tanzu/cartographer/pull/556)
+
+Part of the mechanics requires [Allow Resources to Report Status](https://github.com/vmware-tanzu/cartographer/pull/738)
+or a similar mechanism.
+
+# Motivation
+
+[motivation]: #motivation
+
+## Why should we do this?
+
+It is desirable to know which inputs yielded an output, as this knowledge is central to tracing the preceding artifact
+inputs which ultimately result in said output.
+
+While the necessary information to do so may exist on the stamped resources, and thus could be inferred from inspecting
+the resource, Cartographer does not provide an agnostic means to leverage this. This means users must delve into these
+resources and make these inferences themselves. Doing so requires more explicit knowledge of exploring resources, is
+laborious, and error-prone.
+
+While this RFC does not directly propose UX changes to enable this, it describes mechanisms that enable Cartographer to
+internally recognize which inputs yield a given output, which can be used as the basis for UX changes.
+
+## What use cases does it support?
+
+The changes proposed here do not surface the correlation between inputs and outputs, but they introduce mechanics which
+will enable us to build features which do so.
+
+An interesting side effect of these mechanics is that Cartographer will not propagate artifacts output by a resource to
+the next resource unless they can be correlated to input provided to the resource by Cartographer. This is desirable
+from a security perspective, and the outcomes of this behavior are explored in detail in the
+[Draft RFC Read Resources Only When In Success State](https://github.com/vmware-tanzu/cartographer/pull/556).
+
+# What it is
+
+[what-it-is]: #what-it-is
+
+New fields are added to templates which allow template authors to specify correlation rules using `observedMatches`,
```

Comment:
Cartographer already has a template with the top level field `observedMatches`: [ClusterDeploymentTemplate](https://cartographer.sh/docs/v0.3.0/reference/template/#clusterdeploymenttemplate).

The proposed `ObservedMatches` works differently than this prior art. The existing implementation matches two fields that exist on the stamped object. The proposal matches fields on the template to fields on the stamped object.

1. Should the name change so that artifact tracing can work with ClusterDeploymentTemplate?
2. Should the name change so that we don't overload a term with two implementations?

Reply:
We can assume that fields in the template exist on the spec of the object. Do we need ObservedMatches to match from template to object? Or can we use the same implementation that exists on ClusterDeploymentTemplate, where users are expected to compare the object's spec to its status?

This implementation has the advantage of allowing users to specify `metadata.generation` must equal `status.observedGeneration` explicitly.

Reply:
Good catch on naming. Actually I noticed the names are not even consistent in my document. I'm going to rename to `correlationRules` for the next iteration.

With regards to the matching implementation, I think matching metadata.generation to status.observedGeneration implicitly in the cases where it is needed helps guarantee mechanics work as desired. There are also stipulations under #spec-changes which also try to limit jsonpath expressions to intended uses.